File: sdk/python/src/usenami_signer/exchanges/kucoin.py
from __future__ import annotations

from typing import Any

import httpx

from usenami_signer.exchanges.base import BaseExchange


class KuCoinExchange(BaseExchange):
    """KuCoin API wrapper with automatic Signer authentication.

    Usage::

        signer = Signer("http://signer-demo.usenami.io:8443")

        # List accounts
        resp = signer.kucoin.get("/api/v1/accounts")
        print(resp.json())

        # Place an order
        resp = signer.kucoin.post("/api/v1/orders", body={
            "clientOid": "my-order-001",
            "side": "buy",
            "symbol": "BTC-USDT",
            "type": "market",
            "size": "0.001",
        })
    """

    exchange_name = "kucoin"
    base_url = "https://api.kucoin.com"

    def get_accounts(self, *, currency: str | None = None) -> httpx.Response:
        params = {"currency": currency} if currency else None
        return self.get("/api/v1/accounts", params=params)

    def get_account(self, account_id: str) -> httpx.Response:
        return self.get(f"/api/v1/accounts/{account_id}")

    def get_balances(self) -> httpx.Response:
        return self.get("/api/v1/accounts")

    def get_orders(
        self,
        *,
        symbol: str | None = None,
        status: str = "active",
        **params: Any,
    ) -> httpx.Response:
        query: dict[str, Any] = {"status": status, **params}
        if symbol:
            query["symbol"] = symbol
        return self.get("/api/v1/orders", params=query)

    def place_order(
        self,
        *,
        client_oid: str,
        side: str,
        symbol: str,
        order_type: str = "market",
        size: str | None = None,
        price: str | None = None,
        funds: str | None = None,
    ) -> httpx.Response:
        body: dict[str, Any] = {
            "clientOid": client_oid,
            "side": side,
            "symbol": symbol,
            "type": order_type,
        }
        if size:
            body["size"] = size
        if price:
            body["price"] = price
        if funds:
            body["funds"] = funds
        return self.post("/api/v1/orders", body=body)

    def cancel_order(self, order_id: str) -> httpx.Response:
        return self.delete(f"/api/v1/orders/{order_id}")

    def get_ticker(self, symbol: str) -> httpx.Response:
        return self.get("/api/v1/market/orderbook/level1", params={"symbol": symbol})

    def get_positions(self) -> httpx.Response:
        return self.get("/api/v1/positions")

File: sdk/typescript/src/exchanges/kucoin.ts
/**
 * KuCoin Futures exchange namespace.
 *
 * Auth: HMAC-SHA256 v2 headers.
 * Signed string: timestamp + method + endpoint + body
 *
 * Reference Python impl: sdk/python/src/usenami_signer/exchanges/kucoin.py
 */

import { BaseExchange } from "./base.js";
import type { SignResult, ExchangeResponse } from "../types.js";

const KUCOIN_FUTURES_BASE_URL = "https://api-futures.kucoin.com";

export class KucoinExchange extends BaseExchange {
  /**
   * GET /api/v1/account-overview — account balances.
   */
  async getAccounts(currency: string = "USDT"): Promise<ExchangeResponse> {
    return this.signAndExecute({
      action: "attest", // KuCoin /accounts is read-only; treat as attestation
      method: "GET",
      path: "/api/v1/account-overview",
      params: { currency },
      exchangeBaseUrl: KUCOIN_FUTURES_BASE_URL,
    });
  }

  /**
   * POST /api/v1/orders — place an order.
   *
   * @example
   *   signer.kucoin.placeOrder({
   *     clientOid: "uuid-here",
   *     symbol: "BTCUSDTM",
   *     side: "buy",
   *     leverage: "5",
   *     size: 1,
   *     type: "limit",
   *     price: "60000",
   *   });
   */
  async placeOrder(body: KucoinOrderBody): Promise<ExchangeResponse> {
    return this.signAndExecute({
      action: "order",
      method: "POST",
      path: "/api/v1/orders",
      body,
      exchangeBaseUrl: KUCOIN_FUTURES_BASE_URL,
    });
  }

  /**
   * DELETE /api/v1/orders/{orderId} — cancel an order.
   */
  async cancelOrder(orderId: string): Promise<ExchangeResponse> {
    return this.signAndExecute({
      action: "cancel",
      method: "DELETE",
      path: `/api/v1/orders/${orderId}`,
      exchangeBaseUrl: KUCOIN_FUTURES_BASE_URL,
    });
  }

  /**
   * KuCoin signature → HTTP headers.
   * KC-API-* family per KuCoin Futures v2 docs.
   */
  protected headersFromSignatureImpl(signature: SignResult["signature"]): Record<string, string> {
    if (!signature || signature.scheme !== "hmac-sha256") {
      throw new Error("KuCoin requires hmac-sha256 signature scheme");
    }
    // Gateway returns signature.value as JSON-encoded header dict
    return JSON.parse(signature.value) as Record<string, string>;
  }
}

export interface KucoinOrderBody {
  clientOid: string;
  symbol: string;
  side: "buy" | "sell";
  leverage: string;
  size: number;
  type: "limit" | "market";
  price?: string;
  stopPrice?: string;
  reduceOnly?: boolean;
}
